### ISSUE TITLE
quit menu superimposed

### ISSUE THREAD

**Issue by legluondunet** [elishacloud/Silent-Hill-2-Enhancements#942]
Hello Elisha,
I have an issue with latest sh2ee version (today's downloads):
when I want to quit the game, quit menu superimposed with main menu, main menu does not disappear to let the place to quit menu :

![image](https://github.com/elishacloud/Silent-Hill-2-Enhancements/assets/325405/ff7c41c3-bba2-4bfe-8cc6-7d245b02f5e5)

Thank you for your help.

my config:
Manjaro 23.1.3 Vulcan
Kernel: x86_64 Linux 6.6.19-1-MANJARO
Proc: AMD Ryzen 9 5900HX
GPU: AMD Radeon RX 6600M
Drivers: Mesa 24.0.2
Resolution: 1920x1080
32 Go RAM
DE: Xfce4

**Comment by elishacloud** [comment]
Can you tell me what build number it is and where you downloaded it from?

**Comment by legluondunet** [comment]
I own cds (DIrector's cut). Where can I see build number?

**Comment by elishacloud** [comment]
You can just upload your log file.

**Comment by legluondunet** [comment]
This log file?

[d3d8.log](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/14832489/d3d8.log)

**Comment by elishacloud** [comment]
Try changing the front buffer capture mode.  There is GDI and DirectX.  Try them both and let me know if either of them work for you or not.

![image](https://github.com/elishacloud/Silent-Hill-2-Enhancements/assets/25625456/1a87bde4-1104-4cfa-a7e8-c059e075f6e4)

**Comment by legluondunet** [comment]
Results:
- directx: issue present
- GDI: issue not present:
![image](https://github.com/elishacloud/Silent-Hill-2-Enhancements/assets/325405/c1b77965-a840-45a2-b9ad-f803aa87cd40)
Full black background menu is a normal behavior?

**Comment by elishacloud** [comment]
That is normal for Linux, for some reason.  Linux handles GDI differently.  But not sure why DirectX is not working here.  Either way there is an issue with how the wrapper captures the front buffer.

**Comment by legluondunet** [comment]
I made some tests:
- try different WIne versions: issue still present
- DXVK off: no issue
- DXVK 2.3.1:  issue still present
- DXVK 2.3:  issue still present
- DXVK 2.2:  issue still present
- DXVK 2.1 and less: no issue
![image](https://github.com/elishacloud/Silent-Hill-2-Enhancements/assets/325405/6c4ed823-afe5-4698-b1e5-73cb3a7d3ee2)

Do you have a contact at the DXVK devs team?

**Comment by elishacloud** [comment]
I don't.  But you can open an issue on their GitHub page:  https://github.com/doitsujin/dxvk

**Comment by doitsujin** [comment]
So what exactly does this actually do with the "DirectX" option? Does this just call GetFrontBufferData on a swap chain before actually rendering the menu or something?

GDI is not expected to work since we don't want to do presentation on the CPU in DXVK.

**Comment by elishacloud** [comment]
> Does this just call GetFrontBufferData on a swap chain before actually rendering the menu or something?

Yes, that is correct.  You can see all the code [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/blob/48c7085c023c65b7ca574e461f98a007ff7b4cd9/Wrappers/d3d8/IDirect3DDevice8.cpp#L2779-L2942).  We call that command after the Present() before any other call happens, see [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/blob/48c7085c023c65b7ca574e461f98a007ff7b4cd9/Wrappers/d3d8/IDirect3DDevice8.cpp#L1066-L1108).

Note: the game was created in Direct3D8 and we use [d3d8to9](url) to convert it to Direct3D9.  Also, we use an old build of ReShade to allow us to add shaders, see applicable code [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/blob/48c7085c023c65b7ca574e461f98a007ff7b4cd9/Wrappers/d3d9/IDirect3DDevice9.cpp#L415-L424).  This means that the call may be going through a swap chain.

Let me know if you need any other information here.

**Comment by JoonasPitk** [comment]
The only way I've ever managed to mitigate the issue fully was to set the Screen Mode as Fullscreen.

**Comment by K0bin** [comment]
> We call that command after the Present() before any other call happens, see [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/blob/48c7085c023c65b7ca574e461f98a007ff7b4cd9/Wrappers/d3d8/IDirect3DDevice8.cpp#L1066-L1108).

@elishacloud 

Just looking at the D3D9 commands that get executed in the end, you pretty much call GetFrontBufferData in the middle of the frame. Looking at the code you linked, my guess is that it doesn't actually call the underlying Present when GetFrontBufferData gets called. This is correct D3D9 code, it's just bad because we implement GetFrontBufferData by copying from the last backbuffer and the game only uses a single back buffer. This is the first time this has been a problem.

Either way, I have a fix that will do another copy of the backbuffer and keep that around for future GetFrontBufferData calls.
We'll add it as a config option and automatically enable it for sh2pc.exe.

**Comment by elishacloud** [comment]
@K0bin,

> you pretty much call GetFrontBufferData in the middle of the frame.

Yes, we need to call it before the Present so that it gets the last frame not the current one.  Also, with the latest build we no longer call GetFrontBufferData there.  [Here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/blob/582e7131c5100292aa3f6037bcaeef37ed1a5aaa/Wrappers/d3d8/IDirect3DDevice8.cpp#L1751-L1765) is the new code.

> Looking at the code you linked, my guess is that it doesn't actually call the underlying Present when GetFrontBufferData gets called.

Yes, that one frame is skipped.  This skips things like flashes from the shaders.  It is just a single frame that is skipped there.  However, we have a new way to skip the code, see link above.

> it's just bad because we implement GetFrontBufferData by copying from the last backbuffer and the game only uses a single back buffer.

True, but in most cases I create a copy of the frame that is to be copied and I use a double buffer for that.  See [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/blob/582e7131c5100292aa3f6037bcaeef37ed1a5aaa/Wrappers/d3d8/IDirect3DDevice8.cpp#L1571-L1586).

> Either way, I have a fix that will do another copy of the backbuffer and keep that around for future GetFrontBufferData calls.
> We'll add it as a config option and automatically enable it for sh2pc.exe.

Can you give more details of the fix?

**Comment by K0bin** [comment]
> Can you give more details of the fix?

It's pretty simple. When the game calls present we make a copy of the backbuffer that we keep around in case the game calls GetFrontbufferData. And we only do that for Silent Hill 2 right now.

**Comment by elishacloud** [comment]
> When the game calls present we make a copy of the backbuffer that we keep around in case the game calls GetFrontbufferData.

I thought of this "solution" a long time ago.  It adds significant overhead for every frame when the GetFrontbufferData is only called when you enter a menu.  So this solution will slow the whole game down to fix a menu.  I would rather fix the core issue rather than doing this kind of hack.

Have you tried the latest build?  We fixed some issues in this area with the latest build.  You can download it [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/actions/runs/13618056093).

**Comment by legluondunet** [comment]
I reinstall the game from scratch with latest sh2ee version and I can not reproduce this issue.

**Comment by elishacloud** [comment]
Ok, great.  Then I will close this issue.

**Comment by K0bin** [comment]
We also fixed this on the DXVK side. (Fix is in version 2.6)

**Comment by legluondunet** [comment]
Thank you very much all both for your precious help.
